storage/cloudspanner/map_storage.go

```diff
@@ -0,0 +1,373 @@
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	mapLeafDataTbl = "MapLeafData"
+	// Spanner DB columns:
+	colLeafIndex   = "LeafIndex"
+	colMapRevision = "MapRevision"
+	colLeafHash    = "LeafHash"
+)
+
+var errFinished = errors.New("finished")
+
+// MapStorageOptions is used to configure various parameters of the spanner map storage layer.
+type MapStorageOptions struct {
+	TreeStorageOptions
+}
+
+// NewMapStorage initialises and returns a new MapStorage.
+func NewMapStorage(ctx context.Context, client *spanner.Client) storage.MapStorage {
+	return NewMapStorageWithOpts(client, MapStorageOptions{})
+}
+
+// NewMapStorageWithOpts initialises and returns a new MapStorage using options.
+func NewMapStorageWithOpts(client *spanner.Client, opts MapStorageOptions) storage.MapStorage {
+	ret := &mapStorage{
+		ts:   newTreeStorageWithOpts(client, opts.TreeStorageOptions),
+		opts: opts,
+	}
+
+	return ret
+}
+
+// mapStorage is a CloudSpanner backed trillian.MapStorage implementation.
+type mapStorage struct {
+	// ts provides the merkle-tree level primitives which are built upon by this
+	// mapStorage.
+	ts *treeStorage
+
+	opts MapStorageOptions
+}
+
+func (ms *mapStorage) CheckDatabaseAccessible(ctx context.Context) error {
+	return checkDatabaseAccessible(ctx, ms.ts.client)
+}
+
+func (ms *mapStorage) Snapshot(ctx context.Context) (storage.ReadOnlyMapTX, error) {
+	return nil, ErrNotImplemented
+}
+
+func newMapCache(tree *trillian.Tree) (cache.SubtreeCache, error) {
+	hasher, err := hashers.NewMapHasher(tree.HashStrategy)
+	if err != nil {
+		return cache.SubtreeCache{}, err
+	}
+	return cache.NewMapSubtreeCache(defMapStrata, tree.TreeId, hasher), nil
+}
+
+// Returns a ready-to-use MapTreeTX.
+func (ms *mapStorage) begin(ctx context.Context, tree *trillian.Tree, readonly bool, stx spanRead) (*mapTX, error) {
+	tx, err := ms.ts.begin(ctx, tree, newMapCache, stx)
+	if err != nil {
+		glog.Errorf("failed to treeStorage.begin(treeID=%d): %v", tree.TreeId, err)
+		return nil, err
+	}
+
+	return ms.createMapTX(tx)
+}
+
+func (ms *mapStorage) createMapTX(tx *treeTX) (*mapTX, error) {
+	// Sanity check tx.config
+	if cfg, ok := tx.config.(*spannerpb.MapStorageConfig); !ok || cfg == nil {
+		return nil, fmt.Errorf("unexpected config type for MAP tree %v: %T", tx.treeID, tx.config)
+	}
+
+	return &mapTX{
+		treeTX: tx,
+		ms:     ms,
+	}, nil
+}
+
+func (ms *mapStorage) SnapshotForTree(ctx context.Context, tree *trillian.Tree) (storage.ReadOnlyMapTreeTX, error) {
+	return ms.begin(ctx, tree, true, ms.ts.client.ReadOnlyTransaction())
+}
+
+func (ms *mapStorage) ReadWriteTransaction(ctx context.Context, tree *trillian.Tree, f storage.MapTXFunc) error {
+	return ms.readWriteTransactionNew(ctx, tree, f)
+}
+
+// readWriteTransactionNew uses the spanner.RunInTransaction pattern to execute f.
+func (ms *mapStorage) readWriteTransactionNew(ctx context.Context, tree *trillian.Tree, f storage.MapTXFunc) error {
```

Comment:
`readWriteTransactionNew` looks like a simple pass through for `ReadWriteTransaction`. Should the two be merged?

Reply:
good call, done.